blockprod/src/detail/job_manager.rs

```diff
@@ -0,0 +1,358 @@
+            Id<GenBlock>,
+            UnboundedSender<()>,
+            oneshot::Sender<Result<JobKey, JobManagerError>>,
+        )>();
+
+        let mut job_manager = JobManager {
+            chainstate_handle,
+            get_job_count_sender,
+            new_job_sender,
+            stop_job_sender,
+            shutdown_sender,
+        };
+
+        job_manager.subscribe_to_chainstate(chainstate_sender);
+        job_manager.run(
+            chainstate_receiver,
+            get_job_count_receiver,
+            new_job_receiver,
+            stop_job_receiver,
+            shutdown_receiver,
+        );
+
+        job_manager
+    }
+
+    #[allow(clippy::type_complexity)]
+    fn run(
+        &mut self,
+        mut chainstate_receiver: UnboundedReceiver<Id<GenBlock>>,
+        mut get_job_count_receiver: UnboundedReceiver<oneshot::Sender<usize>>,
+        mut new_job_receiver: UnboundedReceiver<(
+            Id<GenBlock>,
+            UnboundedSender<()>,
+            oneshot::Sender<Result<JobKey, JobManagerError>>,
+        )>,
+        mut stop_job_receiver: UnboundedReceiver<(Option<JobKey>, oneshot::Sender<usize>)>,
+        mut shutdown_receiver: UnboundedReceiver<oneshot::Sender<usize>>,
+    ) {
+        tokio::spawn(async move {
+            let mut jobs: BTreeMap<JobKey, JobHandle> = BTreeMap::new();
+
+            loop {
+                tokio::select! {
+                    event = get_job_count_receiver.recv()
+                        => Self::get_job_count_handler(&jobs, event),
+
+                    event = chainstate_receiver.recv()
+                        => Self::chainstate_handler(&mut jobs, event),
+
+                    event = new_job_receiver.recv()
+                        => Self::new_job_handler(&mut jobs, event),
+
+                    event = stop_job_receiver.recv()
+                        => Self::stop_job_handler(&mut jobs, event),
+
+                    event = shutdown_receiver.recv()
+                        => return Self::shutdown_handler(&mut jobs, event), // Note: the return here
+                }
+            }
+        });
+    }
+
+    fn subscribe_to_chainstate(&mut self, chainstate_sender: UnboundedSender<Id<GenBlock>>) {
+        let chainstate_handle = self.chainstate_handle.clone();
+
+        tokio::spawn(async move {
+            chainstate_handle
+                .call_mut(|this| {
+                    let subscribe_func =
+                        Arc::new(
+                            move |chainstate_event: ChainstateEvent| match chainstate_event {
+                                ChainstateEvent::NewTip(block_id, _) => {
+                                    if let Err(e) = chainstate_sender.send(block_id.into()) {
+                                        log::error!(
+                                            "Chainstate subscriber failed to send new tip: {e:?}"
+                                        )
+                                    }
```

Comment:
We have `log_err_pfx` in the `utils` crate for this. More occurrences below. Your call whether it's an improvement or not.

Reply:
Nice. I've updated everywhere where it worked.